File: sorting_big/python/sort.py
import time
import sys

class Data:
    def __init__(self, id_=0, val=0):
        self.id = long(id_)
        self.value = long(val)

    def __cmp__(self, other):
        return self.value < other.value


class Generator:
    def __init__(self):
        self.a = 48271
        self.c = 0
        self.m = 2147483647
        self.prev = 1

    def gen(self):
        self.prev = (self.prev * self.a + self.c) % self.m
        return self.prev


if len(sys.argv) > 1:
    RecordsNum = int(sys.argv[1])
else:
    RecordsNum = 20000000

print ('Building {0} records...'.format(RecordsNum))
start = time.time()
d = [None] * RecordsNum
gen = Generator()

for i in xrange(0, RecordsNum):
    v = gen.gen()
    d[i] = Data(v + 10, v)

print ('Time elapsed: {0} ms'.format(int(1000*(time.time() - start))))

print ('Sorting {0} records...'.format(RecordsNum))
start = time.time()
d.sort(key=lambda item: item.value)
print ('Time elapsed: {0} ms'.format(int(1000*(time.time() - start))))

print ('Checking whether the collection is sorted...')
start = time.time()
for i in xrange(0, RecordsNum - 1):
    if d[i].value > d[i + 1].value:
        print ('Not sorted!')
        break
print ('Time elapsed: {0} ms'.format(int(1000*(time.time() - start))))

File: sorting_big/cpp/sort.cpp
#include <algorithm>
#include <cstdint>
#include <random>
#include <chrono>
#include <iostream>
#include <assert.h>

class Timer
{
public:
    Timer() : beg{clock::now()}
    { }
    ~Timer() {
        auto diff = clock::now() - beg;
        std::cout << "Time elapsed: " 
            << std::chrono::duration_cast<std::chrono::seconds>(diff).count() << "s\n";
    }
private:
    using clock = std::chrono::high_resolution_clock;
    clock::time_point beg;
};

struct Data
{
    uint64_t id;
    uint64_t value;
};

template<typename RndEngine>
struct DataGen
{
    RndEngine e1, e2;
    Data operator()() {
        return {e1(), e2()};
    }
};

std::vector<Data> Generate(size_t amount)
{
    std::cout << "Generating " << amount << " records...\n";
    Timer t;
    DataGen<std::minstd_rand> generator;
    std::vector<Data> d(amount);
    std::generate(std::begin(d), std::end(d), generator);
    return d;
}

constexpr size_t RecordsNum = 1000000000;

int main(int, char**)
{
    auto cmp =  [](const Data& d1, const Data& d2) {
        return d1.value < d2.value;
    };

    auto data = Generate(RecordsNum);

    assert(!std::is_sorted(std::begin(data), std::end(data), cmp));

    {
        std::cout << "Sorting " << data.size() << " elements...\n";
        Timer t;
        std::sort(std::begin(data), std::end(data), cmp);
    }

    {
        std::cout << "Checking whether the collection is sorted...\n";
        Timer t;
        bool sorted = std::is_sorted(std::begin(data), std::end(data), cmp);
        if (!sorted) {
            std::cout << "Collection is not sorted!\n";
        }
    }
    return 0;
}
